### DISCUSSION TITLE
Intrinsic States & Actions on States

### DISCUSSION THREAD

**Discussion by meganindya** [sugarlabs/musicblocks-v4#9] [Ideas]
`instruction sequence` will be available from _block-tree_.

<hr />

# Music

#### State

- `time signature` (duration, measure)

#### Action

- `dot` (works on `time signature`) || what about multiple `note` blocks in a `dot`?
- `tie` (works on `time signature` of 2 same notes) || what about 2 mismatched notes, and all other wrong cases?
- `multiply note value` (works on `time signature`)
- `skip notes` (works on the `instruction sequence`)
- `swing` (works on `time signature` of 2 notes) || what about 1 note or more than 2 notes?

#### State

- `measure` (beat count, beat equivalent note)
- `measure count` (value)
- `current measure` (value)
- `beat factor` (value) || how is this calculated?
- `notes played list` (note values)

#### Action

- `pickup` ( ? ) || what does this do?
- `no clock` ( ? ) || what does this do?
- `note counter` (works on `instruction sequence`)
- `sum note values` (works on `instruction sequence`)

#### State

- `beats per minute` (value, beat equivalent note)
- `beat count` (value)
- `beat-based callback` (beat type, event)

#### Action

- `on every beat do` (dispatches `event`)
- `on strong beat do` (dispatches `event`)
- `on weak beat do` (dispatches `event`)
- `on every note do` (dispatches `event`)

#### State

- `pitch signature` (type, representation values)
- `last pitch played` (type, representation value)
- `current pitch` (type, representation value)

#### Action

- `override accidental` (works on `pitch signature`)
- `scalar transpose` (works on `last pitch played`)
- `semi-tone transpose` (works on `last played pitch`)
- `register` (works on `pitch signature`)
- `invert` (works on `pitch signature`)
- `set pitch number offset` (works on `pitch signature`)
- _utilities on static values_
- `scalar step up` (works on `key` and `mode`)
- `scalar step down` (works on `key` and `mode`)
- **`play pitch`**

#### State

- `key` (mode note pitch, mode)
- `current key` (mode note pitch)
- `moveable do` (yes or no)

#### State

- `mode list` (name, pitch numbers)
- `current mode` (name)
- `intervals list` (name, pitch numbers)

#### Action

- `define mode` (static utility)
- `scalar interval` (works on `mode` and `pitch signature`)
- `semi-tone interval` (works on `mode` and `pitch signature`)
- `scalar interval measure` (works on `instruction sequence` and `mode`)
- `semi-tone interval measure` (works on `instruction sequence` and `mode`)
- `doubly` (works on `intervals list`)
- `set temperament` (works on `pitch signature` and `mode`)

#### State

- `instrument` (name)
- `oscillator` (modulation type)

#### Action

- `FM Synth` (works on `oscillator`)
- `AM Synth` (works on `oscillator`)
- `Duo Synth` (works on `oscillator`)

#### State

- `effects list` (name)
- `partials list` (weight)

#### Action

- `vibrato` (`tone.js` utility that works on synth corresponding to `instrument` and `effects list`)
- `chorus` (`tone.js` utility that works on synth corresponding to `instrument` and `effects list`)
- `phaser` (`tone.js` utility that works on synth corresponding to `instrument` and `effects list`)
- `tremolo` (`tone.js` utility that works on synth corresponding to `instrument` and `effects list`)
- `distortion` (`tone.js` utility that works on synth corresponding to `instrument` and `effects list`)
- `harmonic` (`tone.js` utility that works on synth corresponding to `instrument`, `effects list`, and `partials list`)
- `weighted partials` ( ? ) || what does this do?

#### State

- `volume` (100-scale value for all instruments/drums)
- `instrument/drum-volume map` (name: value)

#### Action

- `crescendo` (works on `volume`)
- `decrescendo` (works on `volume`)
- `set panning` ( ? ) || what does this do?
- `set synth volume` (works on `instrument/drum-volume map`)
- `set drum volume` (works on `instrument/drum-volume map`)

#### State

- `current drum` (name)

#### Action

- **`play drum`**

<hr />

# Logic

Most only maintain a dynamic state and work on the `instruction sequence`.

#### State

- `action map` (name, event)
- `boxes` (name, value)
- `heaps` (name, data structure)
- `dictionaries` (name, data structure)

#### Action

- `dispatch action` (dispatches an `event` from `action map`)
- _other dispatch_ `event`s.
- _operations on_ `boxes`
- _static utilities_
- _operations on_ `heaps`
- _operations on_ `dictionaries`

<hr />

# Art & Interaction

#### State

- `position` (x-coordinate, y-coordinate)
- `heading` (angle)
- `spline constraints` (values)

#### Action

- `draw on canvas`

#### State

- `color` (100-scale value)
- `lightness` (100-scale value)
- `saturation` (100-scale value)
- `hue` (100-scale value)
- `alpha` (100-scale value)
- `pen-size` (100-scale value)
- `pen state` (up or down)
- `fill state` (yes or no)
- `line type` (solid or hollow)

#### Action

- **`effect canvas`**

#### State

- `CanvasContext properties`

#### Actions

- `file actions`
- `CanvasContext` _utilities_

#### State

- `cursor state`
- `cursor position` (x-coordinate, y-coordinate)
- `AudioContext properties`

#### Action

- **`interaction effect`**

#### State

- `mouse name` (name)
- `mouse list` (names)
- `mouse reference lists` (object in _view_)

#### Action

- `dispatch mouse event`

**Comment by walterbender** [comment]
- dot (works on time signature) || what about multiple note blocks in a dot?
They should all get extended.

- tie (works on time signature of 2 same notes) || what about 2 mismatched notes, and all other wrong cases?
The mismatch case would be a glissando (glide). Maybe we can combine these two into one block?

- skip notes (works on the instruction sequence)
There is also duplicate, which shiuld duplicate notes -- similar to repeat but instead of abcdabcd you get aabbccdd

- swing (works on time signature of 2 notes) || what about 1 note or more than 2 notes?
It doesn't mean anything for one note, but is can apply to more than 2 notes.

- beat factor (value) || how is this calculated?
Basically, what beat is the current note in?

- pickup ( ? ) || what does this do?
It impacts measure and beat -- basically it says, don't start counting measure/beat boundaries until the pickup has played. I think we have not fully implemented this -- but it works for Lilypond output.

- no clock ( ? ) || what does this do?
Maybe we can get rid of this concept if we stick to putting notes directly into the tone.js timeline. What we generally do is run a common clock among all of the voices and if a voice falls behind, we shorten a note to catch it up. But sometimes we want to just ignore that and let everything run at whatever pace it can.

**Comment by walterbender** [comment]
- set panning ( ? ) || what does this do?
That is for left-right fade in the output.

**Comment by walterbender** [comment]
- weighted partials ( ? ) || what does this do?
It is used when designing a custom timbre -- how much weight to give each partial in a harmonic.

**Comment by meganindya** [comment]
Added the rest of the palettes too. I think this is sort of an exhaustive list. We are currently storing transitory instructions as states too. The new state model will separate them and thus, we should have only around 50 states as compared to over 160.

**Comment by pikurasa** [comment]
> * swing (works on time signature of 2 notes) || what about 1 note or more than 2 notes?
>       It doesn't mean anything for one note, but is can apply to more than 2 notes.

For one note, I it can add the specified value. For the second, it subtracts the specified value ("swing value").

> The Swing block works on pairs of notes (specified by note value), adding some duration (specified by swing value) to the first note and taking the same amount from the second note.

![Screenshot from 2024-04-17 12-08-52](https://github.com/sugarlabs/musicblocks-v4/assets/13454579/d5d8b7e7-feda-47fb-bd47-25dc3ccf483d)

This being said, I'm not confident that this block is working properly at the moment. Also, we could consider simplifying its design by inferring input, such as: Assuming that the note value of the first note block contained is the note value operated on, thus removing the need for the second "note value" input.
